src/magic_folder/magic_folder.py

```diff
@@ -857,3 +874,92 @@ def add_file(self, path):
         d = defer.Deferred()
         self._queue.put((bytespath, d))
         return d
+
+@attr.s
+class RemoteSnapshotCreator(object):
+    _state_db = attr.ib()
+    _local_author = attr.ib()
+
+    @eliotutil.inline_callbacks
+    def _upload_local_snapshots(self, tahoe_client):
+        """
+        Check the db for uncommitted LocalSnapshots, deserialize them from the on-disk
+        format to LocalSnapshot objects and commit them into the grid.
+
+        :param tahoe_client: a TahoeClient instance
+        """
+
+        # get the mangled paths for the LocalSnapshot objects in the db
+        localsnapshot_names = self._state_db.get_all_localsnapshot_paths()
+
+        # XXX: processing this table should be atomic. i.e. While the upload is
+        # in progress, a new snapshot can be created on a file we already uploaded
+        # but not removed from the db and if it gets removed from the table later,
+        # the new snapshot gets lost. Perhaps this can be solved by storing each
+        # LocalSnapshot in its own row than storing everything in a blob?
+        # https://github.com/LeastAuthority/magic-folder/issues/197
+        for name in localsnapshot_names:
+            action = UPLOADER_SERVICE_UPLOAD_LOCAL_SNAPSHOTS(relpath=name)
+            with action:
+                # deserialize into LocalSnapshot
+                snapshot = self._state_db.get_local_snapshot(name, self._local_author)
+
+                # now upload each item in the queue
+                try:
+                    remote_snapshot = yield write_snapshot_to_tahoe(
+                        snapshot,
+                        self._local_author,
+                        tahoe_client,
+                    )
+
+                    # At this point, remote snapshot creation successful for
+                    # the given relpath. Remove the LocalSnapshot from the db.
+                    yield self._state_db.delete_localsnapshot(name)
+
+                    # store the remote snapshot capability in the db.
+                    yield self._state_db.store_remotesnapshot(name, remote_snapshot)
+
+                except NoServersError:
+                    # Unable to reach Tahoe storage nodes because of
+                    # network errors or because the tahoe storage nodes
+                    # are offline. Retry?
+                    # XXX: Perhaps implement exponential backoff for retry?
+                    NO_NETWORK.log()
+                    continue
```

Comment:
Is there a way to write a test to exercise this path?

Reply:
We would need to augment the test code to arrange to throw this error.

So in `test_snapshot` it uses `create_tahoe_client` and `create_tahoe_treq_client` and `create_fake_tahoe_root` to set up the "Tahoe" API we're using in those tests ... so the "lowest level" thing to do would be to augment the fake Tahoe root so that it e.g. always throws an error. We could for example have different kinds of "fake tahoe roots" that do different things or a (more complex) single thing that can be told how to behave. It's probably better to keep things simple with multiple kinds of roots (one kind of root can always wrap another)

To test e.g. an exponential backoff strategy here, the fake root could throw an error on the first request, but not on subsequent requests. So this could be a "wrapping" root that has a `_FakeTahoeRoot` but only calls through to it on the *second* PUT or whatever.